Cluster refactor organized into commits

jhershberg opened the pull request:
WIP FOR EARLY REVIEW: Prepare for multiple clustering solutions. 

This PR is a first step towards implementing "flotilla data shards", that is, redis cluster nodes that do their cluster business according to flotilla clustering instead of the existing, legacy, "OSS clustering." 

In order to do achieve this it was necessary to separate (a) the generic, shared cluster elements such as the fact that there are nodes, primaries, shards, etc., and (b) the specifics of the current OSS clustering implementation such as details stemming from the current cluster's gossip protocol. 

In general, a pragmatic approach was taken leading to some mild oddities in places where we avoided making larger changes. If we want to clean these up in the future, we can and probably should.

The cluster state consists of two primary data structures, clusterState and clusterNode. For the purposes of this refactor these two structures were reworked as follows:
 * clusterState now holds just a few fields and a void* "impl" that each implementation (legacy, and in the future - flotilla) uses for its internal data.
 * clusterNode is now a purely internal construct that is exposed via an opaque "handle" with "accessor" functions such as "clusterNodeIsMaster(clusterNodeHandle node)"

The code has been reorganized into the following files:
- cluster.h - The interface between the rest of Redis and the clustering module
- cluster.c - common clustering functions that will be shared between all cluster implementations + conditional #inclusion of the specific, not-common clustering implementation.
- cluster_legacy.h - Structs and macros used by the legacy clustering implementation
- cluster_legacy.c - implementations of the not-common clustering code
- cluster_flotilla.c/h - TODO: These will have the flotilla clustering implementations

One last disclaimer: There is some 'history' in the commits that comprise this PR that contain some earlier design ideas and various minor fixes.

Commit:
rename cluster.c to cluster_legacy.c

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
move generic cluster stuff to cluster.c, add conditional compilation of cluster implementation

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
move msg and other internal legacy stuff into cluster_legacy

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
make legacy parts of clusterNode private

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
make clusterState fields private

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
hide some more constructs, make flags private

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
A little cleanup of cluster.h

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
fix seg fault

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
Initial move of data path functions back to cluster.c

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
expose interfaces for data path functions

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
move verifyClusterNodeId to cluster.c

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
Reorder API declerations in cluster.h to more friendly groupings

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
Move a few more public functions into cluster.c

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
add switch for makefile

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
move mf_end into private clusterStateInternal

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
move clusterNodeDictType into cluster.c, we need it even in flotilla

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
move cluster slots implementation into cluster.c

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

madolson commented on src/cluster.h:
```diff
@@ -173,257 +51,102 @@ typedef struct {
+int clusterNodeConfirmedReachable(clusterNode* node);
+int clusterNodeIsMaster(clusterNode* node);
+int clusterNodeIsSlave(clusterNode* node);
+int clusterNodeIsFailing(clusterNode* node);
+int clusterNodeTimedOut(clusterNode* node);
```
Might just want a "clusterNodeHealth", since many of these states are overlapping.

jhershberg replied:
What, and like have it return flags or something? Not sure I'm following what you mean here.

madolson replied:
Either return an enum for the allowed states (my preference) or we could return some flags. Redis clients don't care about all these substates, they basically only care about "Can I talk to this node", which is what I expect the high level abstraction to vend.

madolson commented on src/cluster.h:
```diff
@@ -173,257 +51,102 @@ typedef struct {
+int clusterNodeIsSlave(clusterNode* node);
+int clusterNodeIsFailing(clusterNode* node);
+int clusterNodeTimedOut(clusterNode* node);
+int clusterNodeIsMyself(clusterNode* node);
```
If this is a handle, this would make more sense to not be per-clustering implementation, since you should be able to still do myself == node. It would be odd to have the node "change" silently underneath you. Ditto about getMyClusterNode.

jhershberg replied:
Actually, this function is used both in the "shared" cluster code (cluster.c) and in module.c (grrrr). Saying "myself == node" requires holding a static myself. Since IsMyself needs to be determined both within the clustering code and within redis (module.c), it makes more sense to me that the clustering hold the static myself and then answer questions about it.

madolson commented on src/cluster.h:
```diff
@@ -173,257 +51,102 @@ typedef struct {
+int clusterNodeIsFailing(clusterNode* node);
+int clusterNodeTimedOut(clusterNode* node);
+int clusterNodeIsMyself(clusterNode* node);
+int clusterNodeIsNoFailover(clusterNode* node);
```
Not sure why this needs to be exposed.

jhershberg replied:
module.c :-(

madolson replied:
Hmm, that seems like a bad decision, maybe we can deprecate it.

madolson commented on src/cluster.h:
```diff
@@ -173,257 +51,102 @@ typedef struct {
+uint16_t getClusterNodeRedirectPort(clusterNode* node, int use_pport);
+const char *getPreferredEndpoint(clusterNode *n);
+clusterNode *clusterLookupNode(const char *name, int length);
+int isClusterManualFailoverInProgress(void);
```
This feels wrong in the API. I would expect Redis to be driving this, because it knows the detail of the state. I would expect both clustering systems to "initiate" the failover, but they shouldn't be the one monitoring it.

madolson replied:
Thought about this more, and yeah I still think there should be a way to call "start write pause" and "stop write pause" and then get rid of this special manual failover state for cluster. I could send a PR for this now I suppose. Either that, or we should expose an API like, "Stepdown as soon as you can", and the master will wait for a replica to sync and step down.

jhershberg replied:
It does feel wrong but it's required in replication.c

madolson replied:
Ok, to make my position clear, I would like us to build a coherent API, even if that requires some modification to how the code is currently is implemented. If you want to split between refactoring and rewriting, that's OK too, but let's finish up both before we do too much integrating with flotilla.

jhershberg replied:
Sorry for the crossed wires but my response was to your initial comment here, not the second one. I think I did not refresh my browser or something when I wrote it. Anyway, your proposal does sound like a good idea though I would prefer to keep this PR focused on refactoring.

madolson replied:
Cool, we had a brief talk with the core team and agreed that early in Redis 8 was the time for refactoring, so we should be able to get all this stuff done.

madolson commented on src/cluster.h:
```diff
@@ -173,257 +51,102 @@ typedef struct {
+void clusterUpdateMyselfAnnouncedPorts(void);
+void clusterUpdateMyselfIp(void);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: cluster commands
+ */
+void clusterCommand(client *c);
```
I would expect we would split this into the application APIs (readonly, readwrite, slots, shards) and then the admin APIs.

jhershberg replied:
Yeah, I also would like to do that. Of the various cluster sub-commands...
* some make no sense in flotilla
* I generalized the cluster slots implementation and moved it into cluster.c
* There are a few more that have generic implementations that can be shared
* cluster shards is a challenge. It's implemented in a manner that makes it very hard to abstract and make shared.

Here is a cut-and-paste from some notes I took ("return error" means it is N/A in flotilla:

"ADDSLOTS <slot> [<slot> ...]", - RETURN ERROR
"ADDSLOTSRANGE <start slot> <end slot> [<start slot> <end slot> ...]", RETURN ERROR
"BUMPEPOCH", RETURN ERROR
"COUNT-FAILURE-REPORTS <node-id>", RETURN ERROR
"COUNTKEYSINSLOT <slot>", SHARED IMPLEMENT
"DELSLOTS <slot> [<slot> ...]", RETURN ERROR
"DELSLOTSRANGE <start slot> <end slot> [<start slot> <end slot> ...]", RETURN ERROR
"FAILOVER [FORCE|TAKEOVER]", ? SHORT TERM NO FORCE, LONG TERM - YES
"FORGET <node-id>", RETURN ERROR
"GETKEYSINSLOT <slot> <count>", SHARED IMPL
"FLUSHSLOTS", RETURN ERROR
"INFO", IMPLEMENT
"KEYSLOT <key>", SHARED CODE
"MEET <ip> <port> [<bus-port>]", RETURN ERROR
"MYID", IMPLEMENT
"MYSHARDID", IMPLEMENT
"NODES", IMPLEMENT BUT MOST OF THE OUTPUT IS N/A
"    Return cluster configuration seen by node. Output format:",
"    <id> <ip:port@bus-port[,hostname]> <flags> <master> <pings> <pongs> <epoch> <link> <slot> ...",
"REPLICATE <node-id>", RETURN ERROR
"RESET [HARD|SOFT]", RETURN ERROR
"SET-CONFIG-EPOCH <epoch>", RETURN ERROR
"SETSLOT <slot> (IMPORTING <node-id>|MIGRATING <node-id>|STABLE|NODE <node-id>)", RETURN ERROR
"REPLICAS <node-id>", IMPLEMENT
"SAVECONFIG", MAYBE EVENTUALLY
"SLOTS", IMPLEMENT
"SHARDS", IMPLEMENT
"LINKS", RETURN ERROR

In addition we will probably want to add commands.

In short, I totally agree with this comment and I am happy to work on doing this but the PR up to this point represents my attempt to get something ready that we can experiment with. I can do this now, in this PR, or later. What's your take?

jhershberg replied:
I would like to work on this now I think. What do you mean by readonly and readwrite?

madolson replied:
> What do you mean by readonly and readwrite?

My brain was just thinking about the client APIs for the cluster spec, of which `READWRITE` and `READONLY` are there but have nothing to do with the cluster command. 

> "SAVECONFIG", MAYBE EVENTUALLY

I don't know why this would be shared, flotilla shouldn't care?

> FAILOVER

Maybe we can simplify this and just return OK for the arguments we silently ignore. For example AWS used the force argument forever because we didn't really think it through. 

> NODES

We should make a decision if we want to maintain this or not. It's supposed to be an admin command, I'm fine not having it be supported in flotilla and keep pushing clients to adopt slots or shards.

madolson replied:
> I can do this now, in this PR, or later. What's your take?

Made in another comment I see two parts, refactoring and interface implementation, I think they should be separate but not specifically concerned which one is first.

madolson commented on src/cluster.h:
```diff
@@ -173,257 +51,102 @@ typedef struct {
+clusterNode* getMigratingSlotDest(int slot);
+clusterNode* getImportingSlotSource(int slot);
+int isClusterHealthy(void);
+uint16_t getClusterNodeRedirectPort(clusterNode* node, int use_pport);
```
Redirect port isn't any different from normal port. If you look upstream you'll see we've refactored this mostly to be getClusterNodePlaintextPort and getClusterNodeTLSPort. That is the interface we want to expose to flotilla.

jhershberg replied:
So this does not quite allign with what you have upstream as it uses the use_pport param to decide which port to return. That is why I named the function after its usage to try and differentiate the name from just "give me this field." Open to new names but it's neither PlaintextPort or TLSPort.

madolson replied:
Let me try to add more color. The messaging the clusterbus sends around has a pport, because that is how it used to work so that is backwards compatible. The cluster node object has no pport, it just has TLS and plaintext. The decision we made is that we want the higher abstraction, the node, to just expose TLS and plaintext. See https://github.com/redis/redis/blob/unstable/src/cluster.c#L114.

madolson commented on src/cluster.h:
```diff
@@ -173,257 +51,102 @@ typedef struct {
+ */
+void freeThisNodesLink(clusterNode *node);
+void freeNodeInboundLink(clusterNode *node);
+sds clusterGenNodesDescription(client *c, int filter, int use_pport);
```
Does flotilla even want this? Do we want to deprecate the nodes response for it? (I think so)

jhershberg replied:
No but it's in use in debug.c (also in clusterCommand()). There are a bunch things here that are only used in debug.c. One option, if we want to further hide stuff, is that the cluster could expose a clusterDebugCommand. I like that.

madolson replied:
> One option, if we want to further hide stuff, is that the cluster could expose a clusterDebugCommand. I like that

I also like this.

madolson commented on src/cluster.h:
```diff
@@ -173,257 +51,102 @@ typedef struct {
+/*
+ * Functions requiring per-clustering mechanism implementation: Debug and info actions
+ */
+void freeThisNodesLink(clusterNode *node);
+void freeNodeInboundLink(clusterNode *node);
```
Not sure this needs to exist in flotilla.

jhershberg replied:
This definitely does not need to exist but it is required for debug.c. See my comment about clusterDebugCommand

madolson replied:
Ack, agree with earlier suggestion. This is all just for test writing anyways.

madolson commented on src/cluster.h:
```diff
@@ -173,257 +51,102 @@ typedef struct {
+ */
 int clusterSendModuleMessageToTarget(const char *target, uint64_t module_id, uint8_t type, const char *payload, uint32_t len);
 void clusterPropagatePublish(robj *channel, robj *message, int sharded);
+void slotToChannelAdd(sds channel);
+void slotToChannelDel(sds channel);
```
Hmm, let's discuss this a bit. I think we made a mistake with how sharded pubsub was designed, and we might want to fix that for flotilla. I think sharded pubsub should only go to primary and be replicated OR not be propagated. Then flotilla doesn't need to worry about this.

jhershberg replied:
That would be great. Since we do not have a cluster bus in Flotilla I was just procrastinating when I left this in.

madolson replied:
I guess let me make an issue or something about this in the main thread. We can just leave this for now then.

madolson commented on src/cluster.h:
```diff
@@ -173,257 +51,102 @@ typedef struct {
+    void* data;
+} clusterNode;
+
 typedef struct clusterState {
```
I want to think through the benefit of keeping the slot->node and shard->node mapping here, the "Redis topology state". It seems like it might allow a lot more of shared code between the implementations, but at the same time it might make the interface much harder since now the clustering implementations need to have a complex update operation.

jhershberg replied:
Yes. Totally. Also undecided about this.

madolson replied:
I spent like an hour prototyping it, and my conclusion is probably this is fine. Basically:
**Pros:***
- Cluster shards, slots and some other misc commands become much cleaner, since we have all the state to show them.
- The API contract is smaller by a couple of APIs. We can compact a lot of stuff together in the updates, not sure if that is a huge win.

**Cons:**
- We basically need inverse APIs for the node update operations, which are arguably more annoying to orchestrate since you have to understand when to call them. 
- Each implementation still needs it's own map of data, so there is quite a bit of duplication. The data doesn't need to be same I suppose. Duplication of data is worse than duplication of code IMO, lower chance of data drift.

madolson replied:
We might be able to just add some APIs anyway to solve the shards/slots issues, so maybe that is preferable anyways. getShards() and getNodesFrom Shard().

jhershberg replied:
"CLUSTER SHARDS" is already in cluster.c and works with the current exposed API. "CLUSTER SLOTS" on the other hand is very enmeshed in the internal "legacy" implementation, partially for optimization. In the mean time I've left it where it is assuming that we will either create a single more generic implementation later or flotilla will implement it for itself from scratch.

jhershberg replied:
Also, the main parts of the cluster implementation, the parts that serve data, are currently generic implementations in cluster.c. Would it be nice to have more shared code? Definitely, agree with your cost/benefit analysis that it's not worth it in this case.

madolson reviewed (commented):
I thought we were going to do this more like a first party module like how TLS was implemented?

madolson commented on src/cluster.h:
```diff
@@ -173,257 +51,102 @@ typedef struct {
+const char *getPreferredEndpoint(clusterNode *n);
+clusterNode *clusterLookupNode(const char *name, int length);
+int isClusterManualFailoverInProgress(void);
+int getNumSlaves(clusterNode *node);
+clusterNode* getSlave(clusterNode *node, int slave_idx);
```
I kind of dislike that we don't just have an API get to all the nodes, it seems okay to return a list and have the caller free it.

jhershberg replied:
...drum roll...

madolson commented on src/cluster.h:
```diff
@@ -173,257 +51,102 @@ typedef struct {
+/*
+ * Functions requiring per-clustering mechanism implementation: "Accessor" functions
+ */
+char* clusterNodeLastKnownIp(clusterNode *node);
```
From an API mirroring perspective, all of this is "last known" from the classic perspective.

jhershberg replied:
fixed

madolson commented on src/cluster.h:
```diff
@@ -173,257 +51,102 @@ typedef struct {
+sds clusterGenNodesDescription(client *c, int filter, int use_pport);
+sds genClusterInfoString(void);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: "Accessor" functions
+ */
```
From a structure perspective, I would prefer to group all of the "per-clustering" sections together then have sub grouping.

jhershberg replied:
sure :-) Fixed

madolson commented on src/cluster.h:
```diff
@@ -173,257 +51,102 @@ typedef struct {
+ * Functions requiring per-clustering mechanism implementation: Lifecycle events
+ */
 void clusterInit(void);
 void clusterInitListeners(void);
```
```suggestion
void clusterInitLast(void);
```

jhershberg replied:
fixed

madolson commented on src/cluster_legacy.c:
```diff
@@ -0,0 +1,6318 @@
+    int j;
+    int update_config = 0;
+
+    /* Return ASAP if a module disabled cluster redirections. In that case
+     * every master can store keys about every possible hash slot. */
+    if (server.cluster_module_flags & CLUSTER_MODULE_FLAG_NO_REDIRECTION)
+        return C_OK;
+
+    /* If this node is a slave, don't perform the check at all as we
+     * completely depend on the replication stream. */
+    if (nodeIsSlave(myself)) return C_OK;
+
+    /* Make sure we only have keys in DB0. */
+    for (j = 1; j < server.dbnum; j++) {
+        if (dictSize(server.db[j].dict)) return C_ERR;
+    }
```
This seems like generic code, so we probably want to crack this function into two parts.

jhershberg replied:
Do you mean these lines or the whole function...The function could be generalized as well...only clusterSaveConfigOrDie would need to be specialized for the two implementations. Seems like a good idea to me. Thoughts?

madolson commented on src/cluster.c:
```diff
@@ -1,1491 +1,517 @@
+    uint64_t cmd_flags = getCommandFlags(c);
+    /* Cluster is globally down but we got keys? We only serve the request
+     * if it is a read command and when allow_reads_when_down is enabled. */
+    if (!isClusterHealthy()) {
```
I would posit that I think there is utility in this having 3 states:
1. I can serve all traffic. (AFAIK everything is healthy in the cluster so I can serve slots based on my known state)
2. I can serve reads. (I'm not sure about the cluster state, but I know I still own my slots)
3. I should serve no traffic. (I have no idea what is going on, maybe the cluster is completing migrated and I should serve nothing)

I think for example "cluster_allow_reads_when_down" is really a statement about 2, but flotilla might have a different perspective about when a node should be able to serve reads.

jhershberg replied:
This seems reasonable. I mainly worked on a 1-1 translation/reorganization of the concepts into an interface that would allow us two separate (legacy and flotilla) imlementations, not on improvements. Of couse I am open to improvements but would prefer, not die on the hill of, keeping this PR focused as a reorganization, rather than making logical changes to the code.

Commit:
s/clusterNodeLastKnownIp/clusterNodeIp/

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
few more review comments

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

madolson commented on src/module.c:
```diff
@@ -8906,35 +8906,34 @@ int RM_GetClusterNodeInfo(RedisModuleCtx *ctx, const char *id, char *ip, char *m
-        if (node->flags & CLUSTER_NODE_PFAIL) *flags |= REDISMODULE_NODE_PFAIL;
-        if (node->flags & CLUSTER_NODE_FAIL) *flags |= REDISMODULE_NODE_FAIL;
-        if (node->flags & CLUSTER_NODE_NOFAILOVER) *flags |= REDISMODULE_NODE_NOFAILOVER;
+        if (clusterNodeIsMyself(node)) *flags |= REDISMODULE_NODE_MYSELF;
+        if (clusterNodeIsMaster(node)) *flags |= REDISMODULE_NODE_MASTER;
+        if (clusterNodeIsSlave(node)) *flags |= REDISMODULE_NODE_SLAVE;
+        if (clusterNodeTimedOut(node)) *flags |= REDISMODULE_NODE_PFAIL;
+        if (clusterNodeIsFailing(node)) *flags |= REDISMODULE_NODE_FAIL;
+        if (clusterNodeIsNoFailover(node)) *flags |= REDISMODULE_NODE_NOFAILOVER;
```
Actually reading this now, yeah I think the per-cluster implementation should generate these flags.

EDIT: So I get Myself, master, slave. I really don't think modules should have introspection into pfail though. I'm not sure about no-failover though and fail though. Fail makes some amount of sense, it corresponds to "unhealthy" basically. There was some discussion that "dead" should really just be not shown in the cluster though.

madolson commented on src/cluster.c:
```diff
@@ -1496,5982 +522,848 @@ int verifyClusterNodeId(const char *name, int length) {
- * to detect timeouts, in order to handle the following case:
- *
- * 1) A client blocks with BLPOP or similar blocking operation.
- * 2) The master migrates the hash slo
```
We could drop clusterNodeIsMyself() if we fix the module.c and then also change getReplOffset() to also take the current replicationGetSlaveOffset().

jhershberg replied:
clusterNodeIsMyself is currently in use in module.c as well. I think it is worth leaving in place the bits required to maintain naive support of the current module.c stuff and then deprecating the whole concept and at that point removing the parts of the cluster API that are not longer required.

madolson replied:
Ack

Commit:
move name+shard_id into internalNode

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
expose only node handles, compiles, passes tests

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
have just one clusterNodeInternal struct, no clusterNode

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
s/clusterNodeInternal/clusterNode/g

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
add handleClusterDebugCommand API, remove two legacy specific APIs

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
split clusterCommand into generic and special implementations

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
split commands.def cluster commands per impl

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
clusterNodeHostname to return char* instead of sds

This makes the interface more natural for clustering impls that
are not native redis.

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
MYID/SHARDID - support no return value

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
feature flags for commands.def generation, flotilla flag

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
remove unused bad def files

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

jhershberg committed:
Update src/cluster.h

Co-authored-by: Madelyn Olson <34459052+madolson@users.noreply.github.com>

Commit:
add failover command to cluster api

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

Commit:
cluster info cmd handling is shared

Signed-off-by: Josh Hershberg <yehoshua@redis.com>

madolson commented on src/cluster_legacy.c:
```diff
@@ -6234,3 +6234,16 @@ int handleDebugClusterCommand(client *c) {
     return 1;
 }
+
+int clusterAllowFailoverCmd(client *c) {
```
Why is this per implementation? I assume we would always direct clients to the cluster failover command?

jhershberg replied:
Correct. It returns true in the flotilla implementation. It's just a simple way to keep the call site generic.
